### PR TITLE
Androapp5035 edit stock

### PR DESCRIPTION
**Pull request by andresmr**
## Description
Please include a summary of the change and include the related jira issue if it exists.

[ jira issue ](https://jira.dhis2.org/browse/

## Solution description
If this PR is a fix include a brief description on how the issue is solved.
## Covered unit test cases
Describe the tests that you ran to verify your changes.
## Where did you test this issue?
- [x] Smartphone Emulator
- [ ] Tablet Emulator
- [x] Smartphone
- [ ] Tablet
## Which Android versions did you test this issue?
- [ ] 4.4
- [ ] 5.X - 6.X
- [ ] 7.X
- [x] 8.X
- [ ] 9.X - 10.X
- [x] Other
## Checklist
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code

### CODE REVIEW

File: stock-managment/src/main/java/org/dhis2/android/rtsm/ui/managestock/components/ManageStockTable.kt [hunk 1]
```diff
@@ -1,55 +1,48 @@
...
-            allowDecimal = false,
-            allowSigned = false
-        )
-    )
+@OptIn(ExperimentalMaterialApi::class)
+fun editingCellValue(
+    editing: Boolean,
+    backdropState: BackdropScaffoldState,
+    scope: CoroutineScope
+) {
```
Comment (andresmr) [inline stock-managment/src/main/java/org/dhis2/android/rtsm/ui/managestock/components/ManageStockTable.kt:43]: I think that you are coupling ManageStockTable with MainContent. You can pass here a lambda as paremeter like: 
```
fun editingCellValue(
    editing: Boolean,
    onEditionStart: () -> Unit
)
    if (editing) {
        onEditionStart.invoke()
    }
}
```
And the in MainContent: 
```
ManageStockTable(manageStockViewModel) {
                        scope.launch { backdropState.conceal() }
```

What do you think?

File: stock-managment/src/main/java/org/dhis2/android/rtsm/ui/managestock/ManageStockViewModel.kt [hunk 2]
```diff
@@ -76,23 +80,31 @@ class ManageStockViewModel @Inject constructor(
...
+
     private val _screenState: MutableLiveData<TableScreenState> = MutableLiveData(
         TableScreenState(emptyList(), false)
     )
     val screenState: LiveData<TableScreenState> = _screenState
 
+    private val _stockItems: MutableLiveData<PagedList<StockItem>> =
+        MutableLiveData<PagedList<StockItem>>()
+
+    private val errors = MutableStateFlow<MutableMap<String, String>>(mutableMapOf())
```
Comment (andresmr) [inline stock-managment/src/main/java/org/dhis2/android/rtsm/ui/managestock/ManageStockViewModel.kt:94]: I think you don't need this: I think error handling can be solved in one part. I'll try to explain with comments. Read below please:

File: stock-managment/src/main/java/org/dhis2/android/rtsm/ui/managestock/ManageStockViewModel.kt [hunk 3]
```diff
@@ -245,6 +266,131 @@ class ManageStockViewModel @Inject constructor(
...
+        return TextInputModel(
+            id = cell.id ?: "",
+            mainLabel = itemName,
+            secondaryLabels = mutableListOf(resources.getString(R.string.quantity)),
+            currentValue = cell.value,
+            keyboardInputType = KeyboardInputType.NumericInput(
+                allowDecimal = false,
+                allowSigned = false
+            ),
+            error = errors.value[cell.id!!]
```
Comment (andresmr) [inline stock-managment/src/main/java/org/dhis2/android/rtsm/ui/managestock/ManageStockViewModel.kt:302]: I think, that we don't need this line neither `error = errors.value[cell.id!!]`

File: stock-managment/src/main/java/org/dhis2/android/rtsm/ui/managestock/ManageStockViewModel.kt [hunk 4]
```diff
@@ -245,6 +266,131 @@ class ManageStockViewModel @Inject constructor(
...
+                                    val data = ruleEffect.data()
+                                    val isValid: Boolean = isValidStockOnHand(data)
+                                    val stockOnHand = if (isValid) data else it.stockOnHand
+                                    addItem(it, cell.value, stockOnHand, !isValid)
+                                    if (!isValid) {
+                                        errors.value[cell.id!!] = resources
+                                            .getString(R.string.stock_on_hand_exceeded_message)
+                                    } else {
+                                        errors.value.remove(cell.id)
+                                    }
```
Comment (andresmr) [inline stock-managment/src/main/java/org/dhis2/android/rtsm/ui/managestock/ManageStockViewModel.kt:335]: I think that we can remove all this code block: 
```
if (!isValid) {
                                        errors.value[cell.id!!] = resources
                                            .getString(R.string.stock_on_hand_exceeded_message)
                                    } else {
                                        errors.value.remove(cell.id)
```
Beacuse we are already storing the error related to the item in line 329:
`addItem(it, cell.value, stockOnHand, !isValid)`
We already have if it is valid or not.

We only need to check it below...

File: stock-managment/src/main/java/org/dhis2/android/rtsm/ui/managestock/ManageStockViewModel.kt [hunk 5]
```diff
@@ -245,6 +266,131 @@ class ManageStockViewModel @Inject constructor(
...
+    ): Map<Int, TableCell> {
+        val stockEntry = getPopulatedEntries().find { it.item.id == cell.id }
+        return tableCells.mapValues { (index, tableCell) ->
+            when (index) {
+                0 -> tableCell.copy(
+                    value = stockEntry?.stockOnHand
+                )
+                else -> tableCell.copy(
+                    value = stockEntry?.qty,
+                    error = errors.value[cell.id]
```
Comment (andresmr) [inline stock-managment/src/main/java/org/dhis2/android/rtsm/ui/managestock/ManageStockViewModel.kt:388]: I think that here we only need to do something like this (maybe there is cleaner way, I didn't think it so much:
```
error = if (stockEntry?.hasError == true) {
                        resources.getString(R.string.stock_on_hand_exceeded_message)
                    } else {
                        null
```
So I think that only with this, we already have error handling working.
What do you think?

**Review by CarlosMacaneta — APPROVED** [review]

**Pull request merged**